File: Python3/401 - 600/589 N-ary Tree Preorder Traversal.py
# 48ms, 83.13%; 14.6MB, 100.00%
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root: Node) -> List[int]:
        if not root: return []
        q, res = [root], []
        while q:
            tmp = q.pop()
            res.append(tmp.val)
            for child in tmp.children[::-1]:
                q.append(child)
        return res

File: 401 - 600/589 N-ary Tree Preorder Traversal.cc
// 56ms, 93.95%; 56.6MB, 10.53%
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    vector<Node *> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node *> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<int> preorder(Node *root) {
        if (!root) return vector<int>();
        vector<Node *> st = {root};
        vector<int> res;
        while (!st.empty()) {
            Node *tmp = st.back();
            st.pop_back();
            res.push_back(tmp->val);
            st.insert(st.end(), tmp->children.rbegin(), tmp->children.rend());
        }
        return res;
    }
};
